### PR TITLE
bulkrax

### PR DESCRIPTION
**Pull request by smithjp**
* https://jira.lib.unc.edu/browse/HYC-981

### CODE REVIEW

File: app/factories/bulkrax/object_factory.rb [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+      end
+    end
+
+    # Override if we need to map the attributes from the parser in
+    # a way that is compatible with how the factory needs them.
+    # override to fix enumeration and formatting of people objects
+    def transform_attributes
+      attrs = attributes.slice(*permitted_attributes)
+      resource = @klass.new
+      attrs.each do |k,v|
```
Comment (bbpennel) [inline app/factories/bulkrax/object_factory.rb:230]: Could you add a little bit of documentation to describe what the cases are covering here?

File: app/factories/bulkrax/object_factory.rb [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        if resource.attributes.keys.member?(k.to_s) && !resource.attributes[k.to_s].respond_to?(:each) && attrs[k].respond_to?(:each)
+          attrs[k] = v.first
+        elsif resource.attributes.keys.member?(k.to_s) && resource.attributes[k.to_s].respond_to?(:each) && !attrs[k].respond_to?(:each)
+          attrs[k] = Array(v)
+        else
+          attrs[k] = v
+        end
+      end
+
+      attrs.each do |k,v|
```
Comment (bbpennel) [inline app/factories/bulkrax/object_factory.rb:240]: Could you add a little bit of documentation to describe what this loop is doing?

File: app/factories/bulkrax/object_factory.rb [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    end
+
+    # Regardless of what the Parser gives us, these are the properties we are prepared to accept.
+    # override to allow '_attributes' properties for people obejcts
+    def permitted_attributes
+      people_types = [:advisors, :arrangers, :composers, :contributors, :creators, :project_directors, :researchers,
+                      :reviewers, :translators]
+      properties = klass.properties.keys.map(&:to_sym)
+      people = properties.map{|p| people_types.include?(p) ? p : nil}.compact
+      properties + people.map{|p| "#{p}_attributes".to_sym} + %i[id edit_users edit_groups read_groups visibility work_members_attributes admin_set_id dcmi_type]
```
Comment (bbpennel) [inline app/factories/bulkrax/object_factory.rb:256]: looks like we are also adding in a few extra fields, worth mentioning in the override comment

File: app/models/concerns/bulkrax/file_factory.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      end
+      @object = object.reload
+      log_deleted_fs(object)
+    end
+
+    def import_files
+      file_paths.map { |path| import_file(path) }
+    end
+
+    # Add file sets to works
```
[the remark this answers is not in the document]
Comment (smithjp) [reply app/models/concerns/bulkrax/file_factory.rb:92]: Works do not have to have filesets in hyrax.  Bulkrax was creating filesets and attaching them to works in a non-standard way which was causing permissions problems, so I changed the object_factory create and update methods and this method to allow filesets to be created by the application the way they normally are and not by sidekiq.

File: app/models/concerns/bulkrax/has_local_processing.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
```
[the remark this answers is not in the document]
Comment (smithjp) [reply app/models/concerns/bulkrax/has_local_processing.rb:1]: This file was generated by the bulkrax gem, but I agree, it should have the override tag.

File: app/parsers/bulkrax/csv_parser.rb [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+      if file? && zip?
+        unzip(parser_fields['import_file_path'], importer_unzip_path)
+        return Dir["#{importer_unzip_path}/*.csv"].first
+      else
+        parser_fields['import_file_path']
+      end
+    end
+
+    # overriding to add array of people types
+    def people_types
```
[the remark this answers is not in the document]
Comment (smithjp) [reply app/parsers/bulkrax/csv_parser.rb:333]: https://jira.lib.unc.edu/browse/HYC-1015

File: config/initializers/hyrax.rb [hunk 7]
```diff
@@ -297,3 +297,8 @@
 Qa::Authorities::Local.register_subauthority('subjects', 'Qa::Authorities::Local::TableBasedAuthority')
 # Qa::Authorities::Local.register_subauthority('languages', 'Qa::Authorities::Local::TableBasedAuthority')
 Qa::Authorities::Local.register_subauthority('genres', 'Qa::Authorities::Local::TableBasedAuthority')
+
+# set bulkrax default work type to first curation_concern if it isn't already set
```
[the remark this answers is not in the document]
Comment (smithjp) [reply config/initializers/hyrax.rb:301]: it should always create a masters paper, but we probably want the general work type to be the default.  i'll change it.

**Pull request merged**